File: 17.24-max-submatrix-lcci/17.24.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import sys
from typing import List

class Solution:
    def getMaxMatrix(self, matrix: List[List[int]]) -> List[int]:
        ans = [0] * 4
        row = len(matrix)
        col = len(matrix[0])
        if row == 0 or col == 0:
            return []
        ans1, ans2 = 0, 0
        maxsum = -sys.maxsize - 1
        for i in range(row):
            arr = [0 for _ in range(col)]
            for j in range(i, row):
                sum1 = 0
                for n in range(col):
                    arr[n] += matrix[j][n]
                    if sum1 > 0:
                        sum1 += arr[n]
                    else:
                        sum1 = arr[n]
                        ans1, ans2 = i, n
                    if maxsum < sum1:
                        maxsum = sum1
                        ans[0] = ans1
                        ans[1] = ans2
                        ans[2] = j
                        ans[3] = n
        return ans

if __name__ == '__main__':
    sol = Solution()
    matrix = [[-1, -2, -9, 6], [8, -9, -3, -6], [2, 9, -7, -6]]
    print(sol.getMaxMatrix(matrix))

    matrix = [[9,-8,1,3,-2],[-3,7,6,-2,4],[6,-4,-4,8,-7]]
    print(sol.getMaxMatrix(matrix))

File: shoope/17.24-max-submatrix-lcci/17.24.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

class Solution {
public:
    vector<int> getMaxMatrix(vector<vector<int>>& matrix) {
        int row = matrix.size();
        int col = row == 0 ? 0 : matrix[0].size();
        if (col == 0) return vector<int>();
        int ans1 = 0, ans2 = 0;
        vector<int> res(4, 0);
        int max = INT_MIN;
        for (int i = 0; i < row; ++i) {
            vector<int> dp(col, 0);
            for (int j = i; j < row; ++j) {
                int sum = 0;
                for (int n = 0; n < col; ++n) {
                    dp[n] += matrix[j][n];
                    if (sum > 0) {
                        sum += dp[n];
                    }
                    else {
                        sum = dp[n];
                        ans1 = i;
                        ans2 = n;
                    }
                    if (sum > max) {
                        max = sum;
                        res[0] = ans1;
                        res[1] = ans2;
                        res[2] = j;
                        res[3] = n; 
                    }
                }
            }
        }
        return res;
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
    vector<vector<int>> matrix1 = {{-1, -2, -9, 6}, {8, -9, -3, -6}, {2, 9, -7, -6}};
    vector<int> res1 = sol.getMaxMatrix(matrix1);
    for (auto& a : res1) {
        cout << a << " ";
    }
    cout << endl;

    vector<vector<int>> matrix2 = {{9,-8,1,3,-2}, {-3,7,6,-2,4}, {6,-4,-4,8,-7}};
    vector<int> res2 = sol.getMaxMatrix(matrix2);
    for (auto& a : res2) {
        cout << a << " ";
    }
    cout << endl;
    return 0;
}